plugins/woocommerce/src/Blocks/Utils/ProductCollectionUtils.php

```diff
@@ -92,4 +99,179 @@ private static function remove_empty_array_recursive( $array ) {
+				$product_taxonomies = get_object_taxonomies( 'product', 'names' );
+				if ( in_array( $taxonomy, $product_taxonomies, true ) ) {
+					$context = 'product-archive';
+				}
+			} elseif ( in_array( $name, array( CartTemplate::SLUG, MiniCartTemplate::SLUG ), true ) ) {
+				$context = 'cart';
+			} elseif ( CheckoutTemplate::SLUG === $name ) {
+				$context = 'checkout';
+			} elseif ( ProductCatalogTemplate::SLUG === $name ) {
+				$context = 'product-catalog';
+			} elseif ( OrderConfirmationTemplate::SLUG === $name ) {
+				$context = 'order-confirmation';
+			}
+		}
+
+		if ( 'page' === $post_type ) {
+			$context = 'page';
+		}
+		if ( 'post' === $post_type ) {
+			$context = 'post';
+		}
+
+		return $context;
+	}
+
+	/**
+	 * Track usage of the Product Collection block within the given blocks.
+	 *
+	 * @param array $blocks     The parsed blocks to check.
+	 * @param bool  $in_single  Whether we are in a single product container (used for keeping state in the recurring process).
+	 *
+	 * @return array Parsed instances of the Product Collection block.
+	 */
+	public static function parse_blocks_track_data( $blocks, $in_single = false ) {
+
+		$instances = array();
+
+		foreach ( $blocks as $block ) {
+			if ( 'woocommerce/product-collection' === $block['blockName'] ) {
+
+				$instances[] = array(
+					'collection'        => $block['attrs']['collection'] ?? 'product-catalog',
+					'in-single-product' => $in_single ? '1' : '0',
+					'filters'           => self::get_query_filters_track_data( $block ),
+				);
+			}
+
+			$local_in_single = $in_single;
+			if ( 'woocommerce/single-product' === $block['blockName'] ) {
+				$local_in_single = true;
+			}
+
+			// Recursive.
+			if ( ! empty( $block['innerBlocks'] ) ) {
+				$instances = array_merge( $instances, self::parse_blocks_track_data( $block['innerBlocks'], $local_in_single ) );
+			}
+		}
+
+		return $instances;
+	}
+
+	/**
+	 * Parse the collection query filters from the query attributes.
+	 *
+	 * @param array $block The parsed block.
+	 * @return array The filters data for tracking.
+	 */
+	public static function get_query_filters_track_data( $block ) {
+
+		if ( ! isset( $block['attrs'] ) ) {
+			return array();
+		}
+
+		$query_attrs = $block['attrs']['query'] ?? array();
+		$filters     = array(
+			'on-sale'      => 0,
+			'stock-status' => 0,
+			'handpicked'   => 0,
+			'keyword'      => 0,
+			'attributes'   => 0,
+			'category'     => 0,
+			'tag'          => 0,
+			'featured'     => 0,
+			'created'      => 0,
+			'price'        => 0,
+		);
+
+		if ( ! empty( $query_attrs['woocommerceOnSale'] ) ) {
+			$filters['on-sale'] = 1;
+		}
+
+		if ( ! empty( $query_attrs['woocommerceStockStatus'] ) ) {
+			$default_values = wc_get_product_stock_status_options();
+			$default_diff   = array_diff( array_keys( $default_values ), $query_attrs['woocommerceStockStatus'] );
+			if ( ! empty( $default_diff ) ) {
+				$filters['stock-status'] = 1;
+			}
+		}
```

Comment:
Does it cover the case when "Out of stock visibility" setting is set to "Hide out of stock items from the catalog" and "Sync with current query" toggle is ON?
In this case, we will need to exclude out-of-stock from default values. 

For more info, You may check [this issue](https://github.com/woocommerce/woocommerce/pull/47537). Maybe this is covered in the PR, but I just wanted to confirm. :slightly_smiling_face:

Reply:
Thanks for catching this, @imanish003! It would have corrupted the stock filter's data! Fixed in https://github.com/woocommerce/woocommerce/pull/46466/commits/7bfdf441f6edc70802a009d8e5f40ed9dd60ffd6